pkg/certificates/bundle.go

```diff
@@ -0,0 +1,99 @@
+package certificates
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+	core "k8s.io/api/core/v1"
+	kubeTypes "k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// CABundleKey is the key in the Kube API Server CA ConfigMap where the CA bundle is located
+	CABundleKey = "ca-bundle.crt"
+	// KubeApiServerOperatorNamespace is the name of the namespace of the ConfigMap that contains the CA for the kubelet
+	// to recognize the kube-apiserver client certificate.
+	KubeApiServerOperatorNamespace = "openshift-kube-apiserver-operator"
+	// KubeAPIServerServingCAConfigMapName is the name of the ConfigMap that contains the CA for the kubelet to
+	// recognize the kube-apiserver client certificate.
+	KubeAPIServerServingCAConfigMapName = "kube-apiserver-to-kubelet-client-ca"
+
+	// configNamespace is the name of the namespace of the ConfigMap that contains the initial kubelet client CA.
+	configNamespace = "openshift-config"
+	// kubeAPIServerInitialCAConfigMapName is the name of the ConfigMap that contains the initial CA certificates
+	// created during the cluster installation, where the initial kubelet CA is valid only for the first year.
+	kubeAPIServerInitialCAConfigMapName = "initial-kube-apiserver-server-ca"
+)
+
+// MergeCertWithCABundle merges the content of the initial bundle with new bundle for the specified subject
+// Taken from https://github.com/openshift/machine-config-operator/blob/1a9f70f333a2287c4a8f2e75cb37b94c7c7b2a20/pkg/operator/sync.go#L939
+func MergeCertWithCABundle(initialBundle, newBundle []byte, subject string) []byte {
+	mergedBytes := []byte{}
+	for len(initialBundle) > 0 {
+		b, next := pem.Decode(initialBundle)
+		if b == nil {
+			break
+		}
+		c, err := x509.ParseCertificate(b.Bytes)
+		if err != nil {
+			// glog.Warningf("Could not parse initial bundle certificate: %v", err)
+			continue
```

Comment:
if we keep hitting in this part, can this become an infinite loop?. The for loop is running as long as `len(initialBundle) > 0 `

Reply:
Correct, sounds like a good case for unit testing.

Reply:
I would like to see unit tests around this functionality

Reply:
Added in https://github.com/openshift/windows-machine-config-operator/pull/1013/commits/6498dc09131d0b61dd01bfef860f36ed60a4e74d